stdlib/entity/resource.lua

```diff
@@ -0,0 +1,84 @@
+--- Resource utilities.
+-- @module Resource
+-- @usage local Resource = require('stdlib/entity/resource')
+
+local Surface = require 'stdlib/area/surface'
+local Tile = require 'stdlib/area/tile'
+local Queue = require 'stdlib/queue'
+
+Resource = {} --luacheck: allow defined top
+
+--- Get the resource entity at the specified position and surface.
+-- Adapted from YARM/resmon.lua -> find_resource_at
+-- @tparam ?|string|LuaSurface surface the surface of the position
+-- @tparam LuaPosition position the position to check
+-- @treturn LuaEntity|nil the resource entity or nil if there is no resource there
+function Resource.get_resource_at(surface, position)
+    local surfaces = Surface.lookup(surface)
+
+    if #surfaces == 0 then
+        return nil
+    end
+
+    local tile_at_position = Tile.from_position(position)
+    local tile_area = Tile.to_area(tile_at_position)
+
+    local resource_at_tile = surface.find_entities_filtered{area=tile_area, type='resource'}
+
+    if #resource_at_tile < 1 then
+        return nil
+    end
+
+    return resource_at_tile[1] -- there should never be another resource at the exact same coordinates
+end
+
+--- From the resource at the given surface and position, return all connected (horizontally, vertically and diagonally) resource entities of the same type.
+-- (For now, just ore patches. Problems arise when a single resource entity spans multiple tiles).
+-- @tparam LuaSurface surface the surface of the position
+-- @tparam LuaPosition position the position to check
+-- @treturn {nil|LuaEntity,...} a table containing the resource entity at the given position and all connected resource entities of the same type or an empty table if there is no resource there
+function Resource.get_resource_patch_at(surface, position)
+    -- get the initial resource tile if there is one at the given position
+    local initial_resource_entity = Resource.get_resource_at(surface, position)
+    if not initial_resource_entity then
+        return {}
+    end
+
+    -- then search around it and expand it to the whole resource patch
+    local resource_patch = {}
+    local resource_type = initial_resource_entity.name
+    local initial_tile = Tile.from_position(initial_resource_entity.position)
+
+    -- do a BFS starting from the initial tile
+    local visited_tiles = {}
+
+    local search_queue = Queue.new()
+    Queue.push_right(search_queue, initial_tile)
+
+    while not Queue.is_empty(search_queue) do
+        local current_tile = Queue.pop_left(search_queue)
+        local current_entity = Resource.get_resource_at(surface, current_tile)
+        local current_tile_index = Tile.get_index(current_tile)
```

Comment:
I have bad news. `Tile.get_index` is only unique for tiles on the same chunk. So the BFS implemented here, if it crosses 2 or more chunks, may not work and abort early. Chunks are 32x32 tiles in Factorio, which are not that big, compared to ore patch dimensions. So it is highly likely this case will occur.

There has been some discussion about unique tile ids on the forums: https://forums.factorio.com/viewtopic.php?f=34&t=41879, but all of the solutions are either slow (using strings) or use chunks for some part of the id system. However - all is not lost. I think a custom tile indexing could work here. The Index computed with `Tile.get_index` only uses 12 of the 32 bits in an integer (0-1023). We could use all 32 bits and achieve an integer index that would be unique as long as the ore patch did not extend more than 65535 tiles in either direction (2^16-1). This would maintain the same performance characteristics here with minimal compromise.

The index function would be:
```lua
bit32.bor(bit32.lshift(bit32.band(current_tile.x, 0xFFFF), 16), bit32.band(current_tile.y, 0xFFFF))
```

So if you replace instances of `Tile.get_index` with that implementation, this should be fixed.